### PR TITLE
Client Direction

### PR DESCRIPTION
**Pull request by clusterfack**
# Adds a way to change client direction

Client dir is a rotational method of byond to view the tiles from a rotated view. It classically breaks literally everything thanks to our current smoothwall code and hotkeys system along with any hardcoded mapping things like sec cell timers.

# How much shit is left

- [x] All window smoothing
- [x] Lattice/catwalk smoothing
- [x] Carpet smoothing
- [x] Fix hotkeys
- [x] Make verbs to change dir
- [x] Unhardcode shuttle smoothing
- [ ] Fully 100% fix shuttle smoothing conditions (diagonals work, kinks work, right angle walls occur too often so they are commented out right now)
- [ ] Fix every multitile object (need sprites probably)
- [x] Wires still have some issues especially the way they are mapped

# What else

- [x] New and smoothed salvage shuttle courtesy of skowron

### CODE REVIEW

**Comment by ComicIronic** [comment]
This is going to be incredible, I almost regret taking the get.

Almost.

Do you have some screens of what it looks like currently?
I wonder if lots of dir changes would cause lag.

**Comment by clusterfack** [comment]
It's just takes a different direction of each icon state based on dir, so it actually doesn't cause lag because its all done by the clients own rendering engine.

**Comment by comma** [comment]
Some floors actually have different icons for different dirs, like ones in vault. Also warning stripes corners are going to break.

**Comment by clusterfack** [comment]
Yeah I'm going to fix those by giving them directional icon states

**Comment by CptWad** [comment]
Oh man you also gotta do shit with the tile painter and other rapid thingywhatsits huh?

**Comment by clusterfack** [comment]
Yeah basically, lots of objects break under the new system and I've been going through and systematically correcting them.

Also just before this is done it will require a LARGE map edit of, at the very least 
-the salvage shuttle on all maps
-sec cell timers which are hardcoded shit

**Comment by CptWad** [comment]
Oh my god please fix the sec cell timers and free mappers from their evil coils

**Comment by clusterfack** [comment]
That's the plan, I want to reduce as much mapping hardcode as I can for this to work. 

Which very likely means I will also rework wall mounted code machine code to be standardized

**Comment by PJB3005** [comment]
\>conflicts already

**Comment by clusterfack** [comment]
Pomf literally just merged something that conflicted with it

**Comment by 9600bauds** [comment]
I voiced my concern in the IRC but I'll voice it here as well, the walls on that salvage shuttle look way too light. Compare to the floor and the airlocks.

**Comment by clusterfack** [comment]
The ground of the salvage shuttle is covered with soot. Also it would be weird if they were the exact same color.

**Comment by 9600bauds** [comment]
The ground is even DARKER without soot

**Comment by clusterfack** [comment]
I'm not sure what to tell you the shading is almost exactly the same as what it is right now.

**Comment by PJB3005** [comment]
IMO the shading's slightly brighter too, or atleast a lot less complex of a sprite, IMO the old salvage shuttle sprites are just tinted shuttle sprites, these ones don't look like that anymore.

**Comment by clusterfack** [comment]
The old salvage shuttle sprites are just squares, all of them with no smoothing, so this is actually infinitely more complex.

**Comment by PJB3005** [comment]
Well, the old salvage shuttle sprite was just the 1 tile shuttle sprite but darker, these new ones aren't the shuttle sprites but darker.

Also, comparing ALL the smoothwall sprites to the old non smoothwalling sprite isn't really complexity, the old sprite was a LOT more complex than these new ones

**Comment by Lokiamis** [comment]
Pretty ship, friend.

**Comment by NigglyWiggly** [comment]
>Smoothwalled salvage shuttle

I don't know how I feel about this

**Comment by Midaychi** [comment]
What if every client got their own randomized client direction on join per round, and kept it till the end of the round?  Bojax1's east might be adamantiumslug's west.

**Comment by ComicIronic** [comment]
It's disorienting as fuck, so I don't think we should do that.

**Comment by Probe1** [comment]
pls finish

**Comment by PJB3005** [comment]
# W H E N
# H
# E
# N

**Comment by Kurfursten** [comment]
I still believe

**Comment by Imagebard** [comment]
Last post 13 days ago

Honk Honk

**Comment by Kurfursten** [comment]
I think you mean "last update six months ago"

**Comment by 9600bauds** [comment]
Someone brought up a fair point. Add this as an effect from Windbreaker. Doesn't matter if it's broken, it's mindbreaker!

**Comment by DrCelt** [comment]
clusterfaaaaaaaack you promised you'd fix it if I got those sprites

**Comment by Zth--** [comment]
Do eet

**Comment by DrCelt** [comment]
What needs fixed?

**Comment by Imagebard** [comment]
Clusterfack, re-add the conflicts label

**Comment by Zth--** [comment]
please finish it

**Comment by 9600bauds** [comment]
never ever

**Comment by clusterfack** [comment]
Ok I'll finish it

**Comment by clusterfack** [comment]
But you know I can't guarantee it'll fix all the issues that changing client.dir has at the current moment and some sprites like the fireplace will look like shit without directional sprites

**Comment by ihadtoregisterforthis** [comment]
Small drawback for such a large gain

**Comment by Exxion** [comment]
As long as Celt fucking with us stops breaking hotkeys I don't give a fuck

File: code/modules/mob/mob_movement.dm [hunk 1]
```diff
@@ -26,44 +26,54 @@
...
+		if(NORTHEAST)
+			swap_hand()
+		if(SOUTHEAST)
+			attack_self()
+		if(SOUTHWEST)
+			if(iscarbon(usr))
+				var/mob/living/carbon/C = usr
+				C.toggle_throw_mode()
+			else
+				usr << "<span class='warning'>This mob type cannot throw items.</span>"
```
Comment (ghost) [inline code/modules/mob/mob_movement.dm:56]: to_chat, isn't it?

File: code/modules/mob/mob_movement.dm [hunk 2]
```diff
@@ -26,44 +26,54 @@
...
+		if(NORTHWEST)
+			if(iscarbon(usr))
+				var/mob/living/carbon/C = usr
+				if(!C.get_active_hand())
+					usr << "<span class='warning'> You have nothing to drop in your hand.</span>"
```
Comment (ghost) [inline code/modules/mob/mob_movement.dm:61]: to_chat

File: code/modules/mob/mob_movement.dm [hunk 3]
```diff
@@ -26,44 +26,54 @@
...
+					return
+				drop_item()
+			else if(isMoMMI(usr))
+				var/mob/living/silicon/robot/mommi/M = usr
+				if(!M.get_active_hand())
+					M << "<span class='warning'> You have nothing to drop or store.</span>"
```
Comment (ghost) [inline code/modules/mob/mob_movement.dm:67]: to_chat

File: code/modules/mob/mob_movement.dm [hunk 4]
```diff
@@ -26,44 +26,54 @@
...
+					return
+				M.uneq_active()
+			else if(isrobot(usr))
+				var/mob/living/silicon/robot/R = usr
+				if(!R.module_active)
+					return
+				R.uneq_active()
+			else
+				usr << "<span class='warning'> This mob type cannot drop items.</span>"
```
Comment (ghost) [inline code/modules/mob/mob_movement.dm:76]: to_chat

**Comment by DrCelt** [comment]
When

**Comment by Brolaire** [comment]
THIS WAS A YEAR AGO

**Comment by Exxion** [comment]
W H E N

**Comment by SkowronX** [comment]
wew

**Comment by Skullyton** [comment]
So, what needs doing with this?

**Comment by Skullyton** [comment]
Welp. Shuttle walls don't make use of dir, but are just changed icons.

How did Unid pull off shuttle rotation?

**Comment by unid15** [comment]
Shuttle rotation works by rotating the icons themselves, instead of changing dir

**Comment by Exxion** [comment]
>Shuttle walls don't use dir

What? Why?

**Comment by 9600bauds** [comment]
Magic and mystery

**Comment by Probe1** [comment]
They don't?

**Comment by DrCelt** [comment]
They don't.

Also, WHEN

**Comment by PJB3005** [comment]
If this becomes over 1 year old without update from cluster I'm getting the closing valids.

**Comment by Zth--** [comment]
we still have tons of /hope/ then

**Comment by Skullyton** [comment]
HWEN?

**Comment by Intigracy** [comment]
One year.

O N E Y E A R

**Comment by PJB3005** [comment]
in 6 days I'm closing this.

**Comment by Exxion** [comment]
>tfw both #5000 and #5001 got abandoned, and both of them would have deserved the get ottherwise

**Comment by Shadowmech88** [comment]
Don't worry guys, Clusterfack still has 6 days to finish!

**Comment by DrCelt** [comment]
w h e n

**Comment by Konater** [comment]
HAPPY BIRTHDAY

**Comment by Skullyton** [comment]
Latest updates screwed client direction stuff more (some things that did turn correctly previously do not, objects that are offset such as the cryo keep the offset in their original direction so it looks weird, etc.), and goonlights just do not like different directions at all, so this may need a re-approach anyway

**Comment by PJB3005** [comment]
Rip.

**Comment by unid15** [comment]
F

**Comment by DrCelt** [comment]
F

**Comment by Iamgoofball** [comment]
F

**Comment by 9600bauds** [comment]
F

**Comment by ihadtoregisterforthis** [comment]
ripepperonis

**Comment by clusterfack** [comment]
Maybe I'll chop it up and PR the pieces later, it constantly outgrew itself and became conflicted.

**Pull request closed without merging**